bsip-0040.md

```diff
@@ -0,0 +1,322 @@
+
+#### `lt, le, gt, ge`
+Stateless assert. Allows explicit type converstion:
+- `int` type: use as is
+- `string` type: use `length` of string as `argument value`
+
+The different asserts read as:
+- `lt`: `Argument value` must be less than `comparative`
+- `le`: `Argument value` must be less than or equal to `comparative`
+- `gt`: `Argument value` must be greater than `comparative`
+- `ge`: `Argument value` must be greater than or equal to `comparative`
+
+#### `limit`
+Statefull assert, only `int` type arguments. When the authority is created, `interval_began` is set to `valid_from` from its custom active authority and `max_cumsum` to `0`. Incoming operations are first tried to match all stateless asserts, 
+and if all passes continue with statefull asserts. If `now > interval_began + interval_in_sec`, then set `max_cumsum = 0` and set `interval_began = now`.
+The assert that needs to pass is now `current_cumsum + incoming value <= max_cumsum`. If all `asserts` of this `custom_active_authority` pass, update `current_cumsum = current_cumsum + incoming value`.
+
+#### `limit_monthly`
+Statefull assert, only `int` type arguments. Analogue to `limit`, but `interval_began` is initially set to `month(valid_from)` and set to `month(now)` on update, additionally the time assert is `month(now) >= interval_began + interval_in_months` (include logic for month overflow when year changes).
+
+#### `attribute_assert`
+Stateless assert, only for dictionary type objects. The data list contains restrictions that all must pass, the reference for the `argument` of the child restriction is nested into the attributes of the parent dictionary type object. Allows nesting of `attribute_assert`.
+
+#### `logical_or`
+Stateless assert, only for dictionary type objects. The data is a list of restrictions lists, i.e. 
+`data = [ [restriction 1, restriction 2], [restriction 3, restriction 4], ... ]`. If one of the restrictions sub-lists in data passes as whole, this restriction passes. In above miniexample that would mean if `restriction 1` and `restriction 2` pass, the whole `logical_or` restriction is considered to be passed as well.
+
+### Outline of handling incoming transactions
+ 
+When a signed transaction arrives and before the backend evaluates if all necessary authorities are present through the signatures, do the following:
+- iterate over required accounts and for each account, iterate over all operations within the transactions that require the active authority of this account
+- iterate the `custom_active_authorities` of said account
+- if a `custom_active_authority` is found that matches , remember that and stop iterating the authorities and continue until all operations are checked
+- if the account has a `custom active authority` match for every operation in the transaction that requires it, then grant the `active authority` of said account. If no match is found, treat as if no authority was given
```

Comment:
I think this is misleading, or perhaps I just don't get it.
Current validation logic for reference: https://github.com/bitshares/bitshares-core/blob/2.0.180612/libraries/chain/protocol/transaction.cpp#L249
This is how I understand your proposal:
- iterate over operations:
-- get required authority accounts for operation
-- for each active authority account:
--- if account has matching custom active authority for operation, add this to list of required custom authorities,
--- otherwise add account to list of active authority accounts
- Verify active/owner/other accounts as in current logic
- on top of that, verify new list of custom active authorities

This means for example that if Alice has active authority key K and a custom active authority that allows key K' to sign a transfer of up to 1 BTS to Bob, then 
- Alice can transfer 2 BTS to Bob using K
- Alice can transfer 1 BTS to Charlie using K
- Alice can transfer 1 BTC to Charlie using K
- Alice can sell 1 BTS for whatever using K
- Alice can transfer 1 BTS to Bob, but she **must** use K' for that (K cannot sign)

Note that this new logic also applies to proposals. Not sure about the consequences of this.

Reply:
I see. I will try to rewrite the formulation.

Anyways, this is the idea:
 - the existing authority checking is not altered. If active authority is is given and also required, the operation is executed
 - additionally, if the active authority of a required account A is [not found](https://github.com/bitshares/bitshares-core/blob/2.0.180612/libraries/chain/protocol/transaction.cpp#L283) we check if there are signatures in sigs that have a valid custom_active_authority of account A. If such a custom_active_authority is found for every operation that requires active authority of account A, the active authority of account A is [added to sigs](https://github.com/bitshares/bitshares-core/blob/2.0.180612/libraries/chain/protocol/transaction.cpp#L271)

For your example:

- Alice can transfer 2 BTS to Bob using K
- Alice can transfer 1 BTS to Charlie using K
- Alice can transfer 1 BTC to Charlie using K
- Alice can sell 1 BTS for whatever using K
- Alice can transfer up to 1 BTS to Bob using K
- Anyone can transfer up to 1 BTS from Alice to Bob, as long as transaction is sigend with K'

Reply:
I think it is a bad idea to add active authority to the sigs because that could have unforeseeable side effects (for example, account A could be part of a multisig authority of account B - adding active authority A might authorize an operation on account B that would otherwise not have been granted).

Reply:
Custom active authority is designed to grant active authority. So if a key from a active authority is used on a matching operation it should grant the active authority. If it is a multi-sig is irrelevant.

**Example**

Account A has Account B and C as multisig. A has custom active authority that grants key K transfer priviliges. B has a custom active authority that grants key L transfer priviliges.

The transaction contains a transfer from A to D.
- signed by B and C: Accepted
- signed by L and C: Denied. The custom active authority of B does not match, only applies for transfers with B as sender 
- signed by K: Accepted, basically bypassing multisig. This is intended, as the multisig needs to approve the installment of the custom active authority in the first place

Reply:
> - signed by L and C: Denied. The custom active authority of B does not match, only applies for transfers with B as sender

I actually made a mistake here. The requirement is not that B is the sender, but that B is a required account. In the backend, in above example. Is A the only required account in a multi-sig setup? Or are B and C also required accounts?

Reply:
Good question. IIRC A is in required active, and the signature checking logic will recurse and check for presence of B and C authorities.

I had a slightly different scenario in mind: suppose Alice has a custom active authority with key K for transfers to Charlie. Bob has Alice as his active authority account.
A proposal containing two operations
1. transfer 1 BTS from Alice to Charlie
2. transfer 1000 BTS from Bob to owner of K
Proposal is approved by K. According to your pseudocode this means that active authority for Alice will be granted (operation 1 requires active authority of Alice, which is fulfilled by custom active authority). Presence of Alice's active authority is sufficient to authorize operation 2, because Alice is authoritative for Bob. Unintended consequence is that owner of K has control over Bob's account. That's why I think "granting active authority" is a bad idea. Instead, K should only be sufficient to authorize operation 1, and Alice's immediate active authority should be required for operation 2.

Reply:
I see. That case should be excluded, as the custom active authority *should* only work directly for required_active accounts, no recursion.

In your example that means the active authority of A is granted *only* to match against required_active, but not to check if A may grant any other active authority as well.

Reply:
I changed the wording from "granting" to "behave as if present for matched operation"

Reply:
+1